src/saml/logout.ts

```diff
@@ -0,0 +1,114 @@
+
+export interface NameID {
+  value: string | null;
+  format: string | null;
+}
+
+const promiseWithNameId = async (nameid: Node): Promise<NameID> => {
+  const format = xpath.selectAttributes(nameid, "@Format");
+  return {
+    value: nameid.textContent,
+    format: format && format[0] && format[0].nodeValue,
+  };
+};
+
+const getNameId = async (doc: Node, decryptionPvk: string | Buffer | null): Promise<NameID> => {
+  const nameIds = xpath.selectElements(
+    doc,
+    "/*[local-name()='LogoutRequest']/*[local-name()='NameID']"
+  );
+  const encryptedIds = xpath.selectElements(
+    doc,
+    "/*[local-name()='LogoutRequest']/*[local-name()='EncryptedID']"
+  );
+
+  if (nameIds.length + encryptedIds.length > 1) {
+    throw new Error("Invalid LogoutRequest");
+  }
+  if (nameIds.length === 1) {
+    return promiseWithNameId(nameIds[0]);
+  }
+  if (encryptedIds.length === 1) {
+    decryptionPvk = assertRequired(
+      decryptionPvk,
+      "No decryption key found getting name ID for encrypted SAML response"
+    );
+
+    const encryptedDatas = xpath.selectElements(
+      encryptedIds[0],
+      "./*[local-name()='EncryptedData']"
+    );
+
+    if (encryptedDatas.length !== 1) {
+      throw new Error("Invalid LogoutRequest");
+    }
+    const encryptedDataXml = encryptedDatas[0].toString();
+
+    const decryptedXml = await decryptXml(encryptedDataXml, decryptionPvk);
+    const decryptedDoc = parseDomFromString(decryptedXml);
+    const decryptedIds = xpath.selectElements(decryptedDoc, "/*[local-name()='NameID']");
+    if (decryptedIds.length !== 1) {
+      throw new Error("Invalid EncryptedAssertion content");
+    }
+    return await promiseWithNameId(decryptedIds[0]);
+  }
+  throw new Error("Missing SAML NameID");
+};
+
+export const processValidlySignedPostRequest = async (
+  doc: XMLOutput,
+  dom: Document,
+  decryptionPvk: string | Buffer | null
+): Promise<{ profile: Profile; loggedOut: true }> => {
+  const request = doc.LogoutRequest;
+  if (request) {
+    const profile = {} as Profile;
+    if (request.$.ID) {
+      profile.ID = request.$.ID;
+    } else {
+      throw new Error("Missing SAML LogoutRequest ID");
+    }
+    const issuer = request.Issuer;
+    if (issuer && issuer[0]._) {
+      profile.issuer = issuer[0]._;
+    } else {
+      throw new Error("Missing SAML issuer");
+    }
+    const nameID = await getNameId(dom, decryptionPvk);
+    if (nameID && nameID.value) {
+      profile.nameID = nameID.value;
+      if (nameID.format) {
+        profile.nameIDFormat = nameID.format;
+      }
+    } else {
+      throw new Error("Missing SAML NameID");
+    }
+    const sessionIndex = request.SessionIndex;
+    if (sessionIndex) {
+      profile.sessionIndex = sessionIndex[0]._;
+    }
+    return { profile, loggedOut: true };
+  } else {
+    throw new Error("Unknown SAML request message");
+  }
+};
+
+export const processValidlySignedSamlLogout = async (
```

Comment:
I think we have gone both ways on this one. Initially I was against the `Async` suffix, but it was needed for refactor reasons. Then I read some stuff about how having that suffix can make code review easier (very obvious if someone forgot an `await`). I also try to remember that others will be using this library, and there is a strong convention in APIs to use the `async` suffix.

Having said that, anyone using TypeScript wouldn't have this problem. Then again, not all those using this library would be using TypeScript. I think I'm 51% in favor of the suffix since it is very common to have a mix of callback and async functions in JavaScript code, and this makes the difference very obvious to any consumer or reviewer, no matter the IDE.

I think though that we should use the same convention over at `passport-saml` no matter what though. @markstos , do you have any thoughts on this?

Reply:
I'll but the -Async back for now, so we can have this discussion separately

Reply:
FWIW, when I was reviewing the code more thoroughly, it seems that we sometimes use the Async suffix for functions that return Promises and other times don't, so we're internally inconsistent. Thus, no matter what decision we make, we'll have to have some code changes to unify our convention internally.